### ISSUE TITLE
Nccl error running run.py with llama2-7b on 2x4090

### ISSUE THREAD

**Issue by kimbaol** [NVIDIA/TensorRT-LLM#1131]
### System Info

GPU:2xRTX4090
OS:docker(tensorrt-llm make to produce image)
TensorRT-LLM version: 0.9.0.dev2024022000
driver:525.116.03
CUDA Version: 12.3

### Who can help?

@byshiue 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. python convert_checkpoint.py --model_dir /tmp/llama2-7b-hf/ --output_dir /tmp/tllm_checkpoint_tp2 --dtype float16 --tp_size 2
2. trtllm-build --checkpoint_dir /tmp/tllm_checkpoint_tp2/ --output_dir /tmp/2tp_fp16_batch32 --gemm_plugin float16 --max_batch_size 32 --use_custom_all_reduce disable
3. mpirun -n 2 --allow-run-as-root python run.py --engine_dir /tmp/2tp_fp16_batch32/ --max_output_len=50 --tokenizer_dir /tmp/llama2-7b-hf/

### Expected behavior

output tokens

### actual behavior

Failed, NCCL error /home/qiaoxj/TensorRT-LLM/cpp/tensorrt_llm/plugins/ncclPlugin/allreducePlugin.cpp:183 'unknown result code'
Failed, NCCL error /home/qiaoxj/TensorRT-LLM/cpp/tensorrt_llm/plugins/ncclPlugin/allreducePlugin.cpp:183 'unknown result code'

### additional notes

Also tired on 2xL40S，the same error

**Comment by PerkzZheng** [comment]
This seems to be the same issue as this https://github.com/NVIDIA/TensorRT-LLM/issues/1039.
I will take a look at it together. BTW, can you try NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,ENV,COLL,NET, and save the log as a txt file ? thanks.

**Comment by qiaoxj07** [comment]
> This seems to be the same issue as this #1039. I will take a look at it together. BTW, can you try NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,ENV,COLL,NET, and save the log as a txt file ? thanks.

Thanks for your reply, the log is attached. 
[run.log](https://github.com/NVIDIA/TensorRT-LLM/files/14383779/run.log)

**Comment by aikitoria** [comment]
Hi @PerkzZheng, I have the same issue. Reproduction steps:

- Get a server from vast.ai with 4x 4090 and CUDA >= 12.2
- Launch it using `nvcr.io/nvidia/tritonserver:24.01-trtllm-python-py3` container image. Though I never got to the part of running triton...
- Follow [these instructions](https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#prepare-tensorrt-llm-engines) on the server for cloning and compiling tensorrt-llm up to `pip3 install`
- Follow the instructions for `INT8 KV cache + AWQ` [here](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/llama), except I used a Mistral 7b [fine tune](https://huggingface.co/cognitivecomputations/dolphin-2.6-mistral-7b) instead of Llama 2 7b.
    - `python3 ../quantization/quantize.py --model_dir /workspace/dolphin/ --output_dir /workspace/dolphin-quantized/ --dtype float16 --qformat int4_awq --awq_block_size 128  --calib_size 32 --tp_size 4 --kv_cache_dtype int8`
    - `trtllm-build --checkpoint_dir /workspace/dolphin-quantized/ --output_dir /workspace/dolphin-engine/ --gemm_plugin float16 --max_batch_size 1 --strongly_typed --weight_only_precision int4 --use_custom_all_reduce disable --paged_kv_cache enable --multi_block_mode enable`
- Attempt to run the model with the extra debug parameters you shared here:
    - `NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=INIT,ENV,COLL,NET mpirun --allow-run-as-root -n 4 python3 ../run.py --max_output_len 50 --tokenizer_dir /workspace/dolphin/ --engine_dir /workspace/dolphin-engine/ >> log.txt`

The result: [log.txt](https://github.com/NVIDIA/TensorRT-LLM/files/14394507/log.txt)

Also: The instructions in the documentation don't actually work, and I had to experiment lots to get it to even try to build the engine. For example, it will fail to build AWQ quantized engine with this awesome error message `Calibration failure occurred with no scaling factors detected. This could be due to no int8 calibrator or insufficient custom scales for network layers` unless `--strongly_typed --weight_only_precision int4` is specified which is documented nowhere... the documentation does however tell you to add `--use_weight_only --per_group --weight_only_precision int4_awq --ammo_quant_ckpt_path`, *none* of which are valid options.

**Comment by PerkzZheng** [comment]
@aikitoria thanks for reporting this. we are moving to a unified interface for converting and building engines, so you may see some incompatibility issues, and the team is working on that.

**Comment by aikitoria** [comment]
That's good to know, though the more important issue is the nccl error at the very end, seems like something wrong with 4090s? I see a bunch of people posting it but no solution, some seem to be able to do something on the host but anything they did outside docker is not applicable to these hosting services.

**Comment by PerkzZheng** [comment]
@kimbaol @aikitoria can you try to add --use_py_session when calling run.py ? it will be helpful for me to identify where is the issue happens as I cannot reproduce locally.

**Comment by qiaoxj07** [comment]
> @kimbaol @aikitoria can you try to add --use_py_session when calling run.py ? it will be helpful for me to identify where is the issue happens as I cannot reproduce locally.

The log after adding "--use_py_session" is attached.  @PerkzZheng 
[run2.log](https://github.com/NVIDIA/TensorRT-LLM/files/14400501/run2.log)

**Comment by PerkzZheng** [comment]
@kimbaol looks like the same error. Can you try with the machines with GPUs connected either by PCIE or NVLINK ? I could not find a machine without P2P access, so that might be part of reasons why I cannot reproduce.

**Comment by aikitoria** [comment]
Would you be able to use a cloud server for reproducing? There are plenty of 4x 4090 machines available on vast.ai for example, if there's no local one available. They can launch the prebuilt tritonserver container image.

It still happens if I set NCCL_P2P_DISABLE=1. I can't test using NVLINK as the 4090s can't be connected with that.

I wonder if the P2P not working is a side effect of using these cloud servers and everything would be fine if I had the hardware properly configured locally. Were you able to test on any machines with 4090 GPUs @PerkzZheng ?

**Comment by qiaoxj07** [comment]
> --use_py_session

Hi @PerkzZheng 
I tired it on 4xL40S, still got the same error. 

nvidia-smi topo -m 
![image](https://github.com/NVIDIA/TensorRT-LLM/assets/5410381/2749bd34-4de4-43aa-af27-048afc905e22)

log without "--use_py_session": 
[run3.log](https://github.com/NVIDIA/TensorRT-LLM/files/14403562/run3.log)

log with "--use_py_session": 
[run4.log](https://github.com/NVIDIA/TensorRT-LLM/files/14403565/run4.log)

**Comment by PerkzZheng** [comment]
thanks. I will see if there are anything wrong when initializing the nccl communicators.

**Comment by aikitoria** [comment]
Today I was able to test the exact same build and model on a server running 2x A100 with NVLINK and that worked correctly. So the problem must be somehow related to 4090s not supporting NVLINK?

**Comment by ZihanLiao** [comment]
> Today I was able to test the exact same build and model on a server running 2x A100 with NVLINK and that worked correctly. So the problem must be somehow related to 4090s not supporting NVLINK?

I'm using A800 and got the same error

**Comment by qiaoxj07** [comment]
Hi @PerkzZheng. Maybe this error is caused by I did not use the latest docker build script.

Here is what I find:

1. build docker image using "make -C docker release_build", then execute run.py in docker, no error.
2. build docker image using [build script](https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#option-2-build-via-docker) from tensorrtllm_backend, then execute run.py in docker, got the same error.

I tested on 2xL40S and 2x4090.

**Comment by qiaoxj07** [comment]
> Hi @PerkzZheng. Maybe this error is caused by I did not use the latest docker build script.
> 
> Here is what I find:
> 
> 1. build docker image using "make -C docker release_build", then execute run.py in docker, no error.
> 2. build docker image using [build script](https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#option-2-build-via-docker) from tensorrtllm_backend, then execute run.py in docker, got the same error.
> 
> I tested on 2xL40S and 2x4090.

I compared the two docker images, torch version is different(2.2.0a vs 2.1.2), torch nccl version(2.19.3 vs 2.18.6)

So I tried a quick fix on 2(build from tensorrtllm_backend script), change "torch<=2.2.0a" in requirements.txt to "torch>=2.2.0", which force reinstall torch 2.2.0 (nccl 2.19.3), and then rebuild trtllm whl, then rerun run.py, no error. 

**For those who may encountered the same problem, update torch version to 2.2.0 may help.**

**Comment by PerkzZheng** [comment]
@kimbaol thanks for the experiments. I was just trying to ask you whether the nccl version is different. 
however, I see that the original log you have shared is also based on nccl 2.19.3 + cuda 12.3. Am I missing something here ?
https://github.com/NVIDIA/TensorRT-LLM/issues/1131#issuecomment-1961016818

**Comment by qiaoxj07** [comment]
> @kimbaol thanks for the experiments. I was just trying to ask you whether the nccl version is different. however, I see that the original log you have shared is also based on nccl 2.19.3 + cuda 12.3. Am I missing something here ? [#1131 (comment)](https://github.com/NVIDIA/TensorRT-LLM/issues/1131#issuecomment-1961016818)

@PerkzZheng 
I use "python -c "import torch;print(torch.cuda.nccl.version())"", and it returns (2, 18, 6)

The nccl version from apt-get install libnccl2=${NCCL_VER} libnccl-dev=${NCCL_VER} is 2.19.3
However, if install torch==2.1.2, nvidia-nccl-cu12==2.18.x will installed.

I am not sure whether this error is related to nccl version.

**Comment by qiaoxj07** [comment]
> @kimbaol thanks for the experiments. I was just trying to ask you whether the nccl version is different. however, I see that the original log you have shared is also based on nccl 2.19.3 + cuda 12.3. Am I missing something here ? [#1131 (comment)](https://github.com/NVIDIA/TensorRT-LLM/issues/1131#issuecomment-1961016818)

@PerkzZheng 
You can try to build docker image using [build script](https://github.com/triton-inference-server/tensorrtllm_backend?tab=readme-ov-file#option-2-build-via-docker) from tensorrtllm_backend, and I think you can reproduce the error.

**Comment by PerkzZheng** [comment]
there might be some issues when both 2.18.6 and 2.19.3 exists. @aikitoria @ZihanLiao can you try what @kimbaol suggested, and see if that works on your side ? if so, we should consider moving to torch 2.2 (there was some incompatible issues when upgrading to torch 2.2 internally).

**Comment by ZihanLiao** [comment]
> there might be some issues when both 2.18.6 and 2.19.3 exists. @aikitoria @ZihanLiao can you try what @kimbaol suggested, and see if that works on your side ? if so, we should consider moving to torch 2.2 (there was some incompatible issues when upgrading to torch 2.2 internally).

Have tried building the docker with tensorrtllm_backend build script, but the error still exists.
BTW，when deploying the 72b model with tensor parallelism, the output tokens are unreasonable. This might due to the trt-engine didn't work.
<img width="1058" alt="image" src="https://github.com/NVIDIA/TensorRT-LLM/assets/54339989/ff24f76c-f3bc-403e-904e-9e1945506a86">

**Comment by aikitoria** [comment]
Hmm. I'm unable to test building a custom docker container at the moment, as I am using cloud servers that are themselves running inside docker. But I just tried it again, this time I downloaded the official build of tensorrt-llm from pip rather than attempting to build it myself, and that worked fine!

I used the `nvcr.io/nvidia/tritonserver:24.01-trtllm-python-py3` image and then installed the preview version inside to try and build an engine, as described on the [readme](https://github.com/NVIDIA/TensorRT-LLM?tab=readme-ov-file#installation):
`apt install openmpi-bin libopenmpi-dev`
`pip3 install tensorrt_llm -U --pre --extra-index-url https://pypi.nvidia.com`

After fixing a bunch of random python errors, this allows me to run the engine with 4-way tensor parallelism, despite P2P not being available. It generated perfectly sensible output, so I cannot reproduce the issue above. This installs torch 2.1, so it seems doubtful that upgrading to torch 2.2 is the fix...

To recap my results so far:

Use tritonserver image, build tensorrt-llm inside -> does not work
Use tritonserver image, pip install tensorrt-llm inside -> works fine

That means I can't make any changes to it (since my builds result in broken tensorrt-llm) but now I can continue testing what I originally wanted to try...

**Comment by aikitoria** [comment]
Potentially related: I ran apt upgrade inside the container, which also updated nccl from whatever version came preinstalled. Maybe that's what fixed it.

**Comment by PerkzZheng** [comment]
@aikitoria @ZihanLiao @kimbao can you try removing this line and rebuilding the images ?
https://github.com/NVIDIA/TensorRT-LLM/blob/main/requirements.txt#L1
it seems to be related to the linking issues of nccl.

**Comment by dontloo** [comment]
same issue

the issue seems only on cuda 12.3
using cuda 12.2 is good
